panic on profiles with numeric labels

### Component(s)

receiver/pprof

### What happened?

## Description

Panic reading profiles with numeric labels, like heap profiles.

## Steps to Reproduce

Read a heap profile.   Here's a failing test too:

```diff
diff --git i/pkg/translator/pprof/pprof_to_profiles_test.go w/pkg/translator/pprof/pprof_to_profiles_test.go
index bf4b4a2aba2..727e1200481 100644
--- i/pkg/translator/pprof/pprof_to_profiles_test.go
+++ w/pkg/translator/pprof/pprof_to_profiles_test.go
@@ -185,6 +185,32 @@ func TestStringToLine(t *testing.T) {
 	}
 }
 
+func TestConvertPprofToProfilesWithNumericLabels(t *testing.T) {
+	// Regression test: profiles with numeric labels (e.g. heap profiles with
+	// "bytes" label) caused a panic because []int64 was used as a map key.
+	p := &profile.Profile{
+		SampleType: []*profile.ValueType{
+			{Type: "alloc_objects", Unit: "count"},
+		},
+		PeriodType: &profile.ValueType{Type: "space", Unit: "bytes"},
+		Period:     524288,
+		Sample: []*profile.Sample{
+			{
+				Location: []*profile.Location{{Address: 0x1}},
+				Value:    []int64{1},
+				NumLabel: map[string][]int64{
+					"bytes": {512},
+				},
+			},
+		},
+	}
+
+	require.NotPanics(t, func() {
+		_, err := ConvertPprofToProfiles(p)
+		require.NoError(t, err)
+	})
+}
+
 func TestLinesToString(t *testing.T) {
 	tests := []struct {
 		name     string
```

## Actual Result

```
panic: runtime error: hash of unhashable type []int64 [recovered, repanicked]

goroutine 170 [running]:
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.deferwrap1()
	go.opentelemetry.io/otel/sdk@v1.40.0/trace/span.go:478 +0x28
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0x14001995860, {0x0, 0x0, 0x14000603a40?})
	go.opentelemetry.io/otel/sdk@v1.40.0/trace/span.go:522 +0xa80
panic({0x10f3375a0?, 0x1400195c350?})
	runtime/panic.go:783 +0x120
type:.hash.github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/pprof.attr(0x14001b95598, 0x14000afc1b0?)
	<autogenerated>:1 +0x38
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/pprof.(*lookupTables).getIdxForAttributeWithUnit(0x14001b957a0, {0x14000afc1b0, 0x5}, {0x0, 0x0}, {0x10ea2aea0, 0x14001998168})
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/pprof@v0.146.0/pprof_to_profiles.go:295 +0x1fc
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/pprof.(*lookupTables).getIdxForAttribute(...)
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/pprof@v0.146.0/pprof_to_profiles.go:268
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/pprof.ConvertPprofToProfiles(0x14001aca2c0)
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/pprof@v0.146.0/pprof_to_profiles.go:170 +0xdc4
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/pprofreceiver.(*pprofScraper).scrape(0x14001843c20, {0x1041774cc?, 0x2000?})
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/pprofreceiver@v0.146.0/receiver.go:64 +0x354
go.opentelemetry.io/collector/scraper/xscraper.ScrapeProfilesFunc.ScrapeProfiles(...)
	go.opentelemetry.io/collector/scraper/xscraper@v0.146.2-0.20260219223409-66996adfaaf7/profiles.go:26
go.opentelemetry.io/collector/scraper/scraperhelper/xscraperhelper.wrapObsProfiles.func1({0x110f1b4f8?, 0x14001ad6000?})
	go.opentelemetry.io/collector/scraper/scraperhelper/xscraperhelper@v0.146.2-0.20260219223409-66996adfaaf7/obs_profiles.go:51 +0xd4
go.opentelemetry.io/collector/scraper/xscraper.ScrapeProfilesFunc.ScrapeProfiles(...)
	go.opentelemetry.io/collector/scraper/xscraper@v0.146.2-0.20260219223409-66996adfaaf7/profiles.go:26
go.opentelemetry.io/collector/scraper/scraperhelper/xscraperhelper.scrapeProfiles(0x140014afb60, {0x12281f058, 0x140019b3b20})
	go.opentelemetry.io/collector/scraper/scraperhelper/xscraperhelper@v0.146.2-0.20260219223409-66996adfaaf7/controller.go:126 +0xcc
go.opentelemetry.io/collector/scraper/scraperhelper/xscraperhelper.NewProfilesController.func1(0x2540be400?)
	go.opentelemetry.io/collector/scraper/scraperhelper/xscraperhelper@v0.146.2-0.20260219223409-66996adfaaf7/controller.go:109 +0x20
go.opentelemetry.io/collector/scraper/scraperhelper/internal/controller.(*Controller[...]).startScraping.func1()
	go.opentelemetry.io/collector/scraper/scraperhelper@v0.146.2-0.20260219223409-66996adfaaf7/internal/controller/controller.go:113 +0xe4
sync.(*WaitGroup).Go.func1()
	sync/waitgroup.go:239 +0x4c
created by sync.(*WaitGroup).Go in goroutine 1
	sync/waitgroup.go:237 +0x70
```

### Collector version

21495e76e7b3b39f0e4ecda6ae6440945cbf64ef (near v0.146.0)

### Environment information

## Environment
go 1.25.7

### OpenTelemetry Collector configuration

```yaml
receivers:
  pprof:
    include: /profiles/heap_*.pprof

exporters:
  debug:
    verbosity: detailed

service:
  pipelines:
    profiles:
      receivers: [pprof]
      processors: []
      exporters: [debug]
```

### Log output

```shell

```

### Additional context

_No response_

### Tip

<sub>[React](https://github.blog/news-insights/product-news/add-reactions-to-pull-requests-issues-and-comments/) with 👍 to help prioritize this issue. Please use comments to provide useful context, avoiding `+1` or `me too`, to help us triage it. Learn more [here](https://opentelemetry.io/community/end-user/issue-participation/).</sub>

--- a/.chloggen/issue-46336.yaml
+++ b/.chloggen/issue-46336.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. receiver/filelog)
+component: pkg/translator/pprof
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: fix attribute bug
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [46336]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: [user]

--- a/pkg/translator/pprof/pprof_to_profiles.go
+++ b/pkg/translator/pprof/pprof_to_profiles.go
@@ -4,14 +4,17 @@
 package pprof // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/pprof"
 
 import (
+	"encoding/binary"
 	"errors"
 	"fmt"
+	"math"
 	"reflect"
 	"slices"
 	"strconv"
 	"strings"
 
 	"github.com/google/pprof/profile"
+	"github.com/zeebo/xxh3"
 	"go.opentelemetry.io/collector/pdata/pcommon"
 	"go.opentelemetry.io/collector/pdata/pprofile"
 	semconv "go.opentelemetry.io/otel/semconv/v1.39.0"
@@ -33,7 +36,7 @@ const (
 // attr is a helper struct to build pprofile.ProfilesDictionary.attribute_table.
 type attr struct {
 	keyStrIdx  int32
-	value      any
+	valueHash  uint64
 	unitStrIdx int32
 }
 
@@ -89,6 +92,7 @@ type lookupTables struct {
 	lastStringTableIdx int32
 
 	attributeTable        map[attr]int32
+	attributeHashToValue  map[uint64]any
 	lastAttributeTableIdx int32
 
 	stackTable        map[stackKey]stack
@@ -275,6 +279,50 @@ func (lts *lookupTables) getIdxForString(s string) int32 {
 	return lts.lastStringTableIdx
 }
 
+// getValueHash returns a unique hash for the given value.
+func getValueHash(v any) uint64 {
+	switch val := v.(type) {
+	case string:
+		return xxh3.HashString(val)
+	case int64:
+		var b [8]byte
+		binary.LittleEndian.PutUint64(b[:], uint64(val))
+		return xxh3.Hash(b[:])
+	case float64:
+		var b [8]byte
+		binary.LittleEndian.PutUint64(b[:], math.Float64bits(val))
+		return xxh3.Hash(b[:])
+	case bool:
+		if val {
+			return 1
+		}
+		return 0
+	case []int64:
+		// Cast the slice to a byte slice for zero-allocation hashing if possible,
+		// but for safety/simplicity, we process the slice:
+		h := xxh3.New()
+		var b [8]byte
+		for _, x := range val {
+			binary.LittleEndian.PutUint64(b[:], uint64(x))
+			_, _ = h.Write(b[:])
+		}
+		return h.Sum64()
+	case []float64:
+		h := xxh3.New()
+		var b [8]byte
+		for _, x := range val {
+			binary.LittleEndian.PutUint64(b[:], math.Float64bits(x))
+			_, _ = h.Write(b[:])
+		}
+		return h.Sum64()
+	case []byte:
+		return xxh3.Hash(val)
+	default:
+		// Fallback for unexpected types (nil, etc.)
+		return 0
+	}
+}
+
 // getIdxForAttribute returns the corresponding index for the attribute.
 // If the attribute does not yet exist in the cache, it will be added.
 func (lts *lookupTables) getIdxForAttribute(key string, value any) int32 {
@@ -285,31 +333,25 @@ func (lts *lookupTables) getIdxForAttribute(key string, value any) int32 {
 // If the attribute does not yet exist in the cache, it will be added.
 func (lts *lookupTables) getIdxForAttributeWithUnit(key, unit string, value any) int32 {
 	keyStrIdx := lts.getIdxForString(key)
-
+	valueHash := getValueHash(value)
 	unitStrIdx := noAttrUnit
 	if unit != "" {
 		unitStrIdx = lts.getIdxForString(unit)
 	}
 
-	for attr, idx := range lts.attributeTable {
-		if attr.keyStrIdx != keyStrIdx {
-			continue
-		}
-		if attr.unitStrIdx != unitStrIdx {
-			continue
-		}
-		if !isEqualValue(attr.value, value) {
-			continue
-		}
-		return idx
+	attrKey := attr{
+		keyStrIdx:  keyStrIdx,
+		valueHash:  valueHash,
+		unitStrIdx: unitStrIdx,
+	}
+
+	if id, exists := lts.attributeTable[attrKey]; exists {
+		return id
 	}
 
 	lts.lastAttributeTableIdx++
-	lts.attributeTable[attr{
-		keyStrIdx:  keyStrIdx,
-		value:      value,
-		unitStrIdx: unitStrIdx,
-	}] = lts.lastAttributeTableIdx
+	lts.attributeTable[attrKey] = lts.lastAttributeTableIdx
+	lts.attributeHashToValue[valueHash] = value
 	return lts.lastAttributeTableIdx
 }
 
@@ -425,19 +467,16 @@ func (lts *lookupTables) getIdxForLocation(l *profile.Location) int32 {
 	return lts.lastLocationTableIdx
 }
 
-func isEqualValue(a, b any) bool {
-	return reflect.DeepEqual(a, b)
-}
-
 // initLookupTables returns a supporting elements to construct pprofile.ProfilesDictionary.
 func initLookupTables() lookupTables {
 	lts := lookupTables{
-		mappingTable:   make(map[mm]int32),
-		locationTable:  make(map[loc]int32),
-		functionTable:  make(map[fn]int32),
-		stringTable:    make(map[string]int32),
-		attributeTable: make(map[attr]int32),
-		stackTable:     make(map[stackKey]stack),
+		mappingTable:         make(map[mm]int32),
+		locationTable:        make(map[loc]int32),
+		functionTable:        make(map[fn]int32),
+		stringTable:          make(map[string]int32),
+		attributeTable:       make(map[attr]int32),
+		attributeHashToValue: make(map[uint64]any),
+		stackTable:           make(map[stackKey]stack),
 	}
 
 	// mapping_table[0] must always be zero value (Mapping{}) and present.
@@ -530,8 +569,26 @@ func (lts *lookupTables) dumpLookupTables(dic pprofile.ProfilesDictionary) error
 	}
 	for a, id := range lts.attributeTable {
 		dic.AttributeTable().At(int(id)).SetKeyStrindex(a.keyStrIdx)
-		if err := dic.AttributeTable().At(int(id)).Value().FromRaw(a.value); err != nil {
-			return err
+		attrValue := lts.attributeHashToValue[a.valueHash]
+		if attrValue == nil {
+			if id == 0 {
+				continue
+			}
+			return fmt.Errorf("invalid attribute %#v", a)
+		}
+		if reflect.TypeOf(attrValue).Kind() == reflect.Slice {
+			slice := dic.AttributeTable().At(int(id)).Value().SetEmptySlice()
+			rv := reflect.ValueOf(attrValue)
+			for i := 0; i < rv.Len(); i++ {
+				value := slice.AppendEmpty()
+				if err := value.FromRaw(rv.Index(i).Interface()); err != nil {
+					return err
+				}
+			}
+		} else {
+			if err := dic.AttributeTable().At(int(id)).Value().FromRaw(attrValue); err != nil {
+				return err
+			}
 		}
 		if a.unitStrIdx != noAttrUnit {
 			dic.AttributeTable().At(int(id)).SetUnitStrindex(a.unitStrIdx)

--- a/pkg/translator/pprof/pprof_to_profiles_test.go
+++ b/pkg/translator/pprof/pprof_to_profiles_test.go
@@ -5,9 +5,12 @@ package pprof // import "github.com/open-telemetry/opentelemetry-collector-contr
 
 import (
 	"bytes"
+	"encoding/binary"
 	"errors"
+	"math"
 	"os"
 	"path/filepath"
+	"runtime/pprof"
 	"slices"
 	"strconv"
 	"testing"
@@ -15,6 +18,7 @@ import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/pprof/profile"
 	"github.com/stretchr/testify/require"
+	"github.com/zeebo/xxh3"
 	"go.opentelemetry.io/collector/pdata/pprofile"
 )
 
@@ -229,6 +233,406 @@ func TestLinesToString(t *testing.T) {
 			}
 		})
 	}
+}
+
+func TestLinesToString_SortingBranches(t *testing.T) {
+	tests := []struct {
+		name     string
+		input    []profile.Line
+		expected string
+	}{
+		{
+			name: "line/column/function sorting with nil functions",
+			input: []profile.Line{
+				{Line: 10, Column: 1},
+				{Line: 10, Column: 1},
+				{Function: &profile.Function{ID: 2, Name: "func2"}, Line: 10, Column: 1},
+				{Function: &profile.Function{ID: 1, Name: "func1"}, Line: 10, Column: 1},
+				{Line: 10, Column: 0},
+				{Line: 9, Column: 9},
+			},
+			expected: "-1:9:9;-1:10:0;-1:10:1;-1:10:1;1:10:1;2:10:1",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			lts := initLookupTables()
+			result := lts.linesToString(test.input)
+			require.Equal(t, test.expected, result)
+		})
+	}
+}
+
+func TestGetValueHash(t *testing.T) {
+	var intBytes [8]byte
+	binary.LittleEndian.PutUint64(intBytes[:], uint64(42))
+
+	var floatBytes [8]byte
+	binary.LittleEndian.PutUint64(floatBytes[:], math.Float64bits(3.1415))
+
+	intSliceHasher := xxh3.New()
+	for _, v := range []int64{1, 2, 3} {
+		var b [8]byte
+		binary.LittleEndian.PutUint64(b[:], uint64(v))
+		_, _ = intSliceHasher.Write(b[:])
+	}
+
+	floatSliceHasher := xxh3.New()
+	for _, v := range []float64{1.5, 2.5} {
+		var b [8]byte
+		binary.LittleEndian.PutUint64(b[:], math.Float64bits(v))
+		_, _ = floatSliceHasher.Write(b[:])
+	}
+
+	tests := []struct {
+		name     string
+		value    any
+		expected uint64
+	}{
+		{name: "string", value: "otel", expected: xxh3.HashString("otel")},
+		{name: "int64", value: int64(42), expected: xxh3.Hash(intBytes[:])},
+		{name: "float64", value: 3.1415, expected: xxh3.Hash(floatBytes[:])},
+		{name: "bool true", value: true, expected: 1},
+		{name: "bool false", value: false, expected: 0},
+		{name: "int64 slice", value: []int64{1, 2, 3}, expected: intSliceHasher.Sum64()},
+		{name: "float64 slice", value: []float64{1.5, 2.5}, expected: floatSliceHasher.Sum64()},
+		{name: "byte slice", value: []byte("abc"), expected: xxh3.Hash([]byte("abc"))},
+		{name: "default nil", value: nil, expected: 0},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			require.Equal(t, test.expected, getValueHash(test.value))
+		})
+	}
+
+	require.NotEqual(t, getValueHash("stable"), getValueHash("changed"))
+}
+
+func TestGetIdxForMMAttributes(t *testing.T) {
+	keyForAttrID := func(t *testing.T, lts *lookupTables, id int32) string {
+		t.Helper()
+
+		var keyIdx int32
+		found := false
+		for key, attrID := range lts.attributeTable {
+			if attrID == id {
+				keyIdx = key.keyStrIdx
+				found = true
+				break
+			}
+		}
+		require.True(t, found)
+
+		for s, idx := range lts.stringTable {
+			if idx == keyIdx {
+				return s
+			}
+		}
+
+		t.Fatalf("string index %d not found", keyIdx)
+		return ""
+	}
+
+	tests := []struct {
+		name         string
+		mapping      *profile.Mapping
+		expectedKeys []string
+	}{
+		{
+			name: "build id only",
+			mapping: &profile.Mapping{
+				BuildID: "build-1",
+			},
+			expectedKeys: []string{
+				"process.executable.build_id.gnu",
+			},
+		},
+		{
+			name: "all has flags",
+			mapping: &profile.Mapping{
+				BuildID:         "build-2",
+				HasFunctions:    true,
+				HasFilenames:    true,
+				HasLineNumbers:  true,
+				HasInlineFrames: true,
+			},
+			expectedKeys: []string{
+				"process.executable.build_id.gnu",
+				"pprof.mapping.has_filenames",
+				"pprof.mapping.has_functions",
+				"pprof.mapping.has_inline_frames",
+				"pprof.mapping.has_line_numbers",
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			lts := initLookupTables()
+			ids := lts.getIdxForMMAttributes(test.mapping)
+			require.Len(t, ids, len(test.expectedKeys))
+
+			actualKeys := make([]string, 0, len(ids))
+			for _, id := range ids {
+				actualKeys = append(actualKeys, keyForAttrID(t, &lts, id))
+			}
+
+			slices.Sort(actualKeys)
+			expectedKeys := slices.Clone(test.expectedKeys)
+			slices.Sort(expectedKeys)
+			require.Equal(t, expectedKeys, actualKeys)
+
+			buildIDHash := getValueHash(test.mapping.BuildID)
+			require.Equal(t, test.mapping.BuildID, lts.attributeHashToValue[buildIDHash])
+		})
+	}
+}
+
+func TestGetIdxForLocation(t *testing.T) {
+	lts := initLookupTables()
+
+	fn := &profile.Function{ID: 1, Name: "f", SystemName: "sys", Filename: "f.go", StartLine: 1}
+	mapping := &profile.Mapping{
+		Start:           10,
+		Limit:           20,
+		Offset:          1,
+		File:            "bin",
+		BuildID:         "build-id",
+		HasFunctions:    true,
+		HasFilenames:    true,
+		HasLineNumbers:  true,
+		HasInlineFrames: true,
+	}
+	location := &profile.Location{
+		Address:  0x1000,
+		IsFolded: true,
+		Mapping:  mapping,
+		Line: []profile.Line{
+			{Function: fn, Line: 42, Column: 7},
+		},
+	}
+
+	idx1 := lts.getIdxForLocation(location)
+	idx2 := lts.getIdxForLocation(location)
+
+	require.Equal(t, idx1, idx2)
+	require.NotEqual(t, int32(0), idx1)
+	require.Greater(t, len(lts.mappingTable), 1)
+	require.Greater(t, len(lts.attributeTable), 1)
+}
+
+func TestConvertPprofToProfiles_LabelValidationErrors(t *testing.T) {
+	baseFunction := &profile.Function{ID: 1, Name: "main", SystemName: "main", Filename: "main.go", StartLine: 1}
+	baseMapping := &profile.Mapping{ID: 1, Start: 1, Limit: 2, Offset: 0, File: "bin", BuildID: "abc"}
+	baseLocation := &profile.Location{
+		ID:      1,
+		Address: 0x1000,
+		Mapping: baseMapping,
+		Line:    []profile.Line{{Function: baseFunction, Line: 10, Column: 2}},
+	}
+
+	makeProfile := func(sample *profile.Sample) *profile.Profile {
+		return &profile.Profile{
+			SampleType: []*profile.ValueType{{Type: "cpu", Unit: "ns"}},
+			Sample:     []*profile.Sample{sample},
+			Function:   []*profile.Function{baseFunction},
+			Location:   []*profile.Location{baseLocation},
+			Mapping:    []*profile.Mapping{baseMapping},
+			PeriodType: &profile.ValueType{Type: "cpu", Unit: "ns"},
+			Period:     1,
+		}
+	}
+
+	t.Run("string label with multiple values", func(t *testing.T) {
+		sample := &profile.Sample{
+			Location: []*profile.Location{baseLocation},
+			Value:    []int64{10},
+			Label: map[string][]string{
+				"k": {"v1", "v2"},
+			},
+		}
+
+		_, err := ConvertPprofToProfiles(makeProfile(sample))
+		require.Error(t, err)
+		require.ErrorIs(t, err, errPprofInvalid)
+	})
+
+	t.Run("numeric label with multiple values", func(t *testing.T) {
+		sample := &profile.Sample{
+			Location: []*profile.Location{baseLocation},
+			Value:    []int64{10},
+			NumLabel: map[string][]int64{
+				"n": {1, 2},
+			},
+		}
+
+		_, err := ConvertPprofToProfiles(makeProfile(sample))
+		require.Error(t, err)
+		require.ErrorIs(t, err, errPprofInvalid)
+	})
+}
+
+func TestConvertPprofToProfiles_LabelAndNumUnitPaths(t *testing.T) {
+	fn := &profile.Function{ID: 1, Name: "main", SystemName: "main", Filename: "main.go", StartLine: 1}
+	mapping := &profile.Mapping{
+		ID:             1,
+		Start:          1,
+		Limit:          2,
+		Offset:         0,
+		File:           "bin",
+		BuildID:        "abc",
+		HasFunctions:   true,
+		HasFilenames:   true,
+		HasLineNumbers: true,
+	}
+	location := &profile.Location{
+		ID:      1,
+		Address: 0x1000,
+		Mapping: mapping,
+		Line: []profile.Line{
+			{Function: fn, Line: 10, Column: 2},
+		},
+	}
+
+	prof := &profile.Profile{
+		SampleType: []*profile.ValueType{{Type: "cpu", Unit: "ns"}},
+		Sample: []*profile.Sample{{
+			Location: []*profile.Location{location},
+			Value:    []int64{10},
+			Label: map[string][]string{
+				"label.with.unit": {"v"},
+				"label.no.unit":   {"x"},
+			},
+			NumLabel: map[string][]int64{
+				"num.with.unit": {7},
+				"num.no.unit":   {8},
+			},
+			NumUnit: map[string][]string{
+				"label.with.unit": {"items"},
+				"num.with.unit":   {"ms"},
+			},
+		}},
+		Function:   []*profile.Function{fn},
+		Location:   []*profile.Location{location},
+		Mapping:    []*profile.Mapping{mapping},
+		PeriodType: &profile.ValueType{Type: "cpu", Unit: "ns"},
+		Period:     1,
+		DropFrames: "drop",
+		KeepFrames: "keep",
+		Comments:   []string{"comment"},
+		DocURL:     "https://example.com",
+	}
+
+	out, err := ConvertPprofToProfiles(prof)
+	require.NoError(t, err)
+	require.NotNil(t, out)
+
+	p := out.ResourceProfiles().At(0).ScopeProfiles().At(0).Profiles().At(0)
+	require.Equal(t, 1, p.Samples().Len())
+	require.GreaterOrEqual(t, p.Samples().At(0).AttributeIndices().Len(), 4)
+	require.GreaterOrEqual(t, p.AttributeIndices().Len(), 3)
+}
+
+func TestDumpLookupTables_Errors(t *testing.T) {
+	t.Run("invalid mapping attribute indices", func(t *testing.T) {
+		lts := initLookupTables()
+		lts.mappingTable[mm{attrIdxs: "2;1"}] = 1
+
+		dic := pprofile.NewProfiles().Dictionary()
+		err := lts.dumpLookupTables(dic)
+		require.Error(t, err)
+		require.ErrorIs(t, err, errIdxFormatInvalid)
+	})
+
+	t.Run("invalid location lines format", func(t *testing.T) {
+		lts := initLookupTables()
+		lts.locationTable[loc{lines: "invalid"}] = 1
+
+		dic := pprofile.NewProfiles().Dictionary()
+		err := lts.dumpLookupTables(dic)
+		require.Error(t, err)
+		require.ErrorIs(t, err, errIdxFormatInvalid)
+	})
+
+	t.Run("missing attribute value for non-zero attribute", func(t *testing.T) {
+		lts := initLookupTables()
+		keyIdx := lts.getIdxForString("custom.key")
+		lts.attributeTable[attr{keyStrIdx: keyIdx, valueHash: 12345, unitStrIdx: noAttrUnit}] = 1
+
+		dic := pprofile.NewProfiles().Dictionary()
+		err := lts.dumpLookupTables(dic)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "invalid attribute")
+	})
+}
+
+func TestAllElementsSame(t *testing.T) {
+	require.False(t, allElementsSame([]int{}))
+	require.True(t, allElementsSame([]int{3, 3, 3}))
+	require.False(t, allElementsSame([]int{1, 2, 1}))
+}
+
+func TestGetAttributeString(t *testing.T) {
+	dic := pprofile.NewProfiles().Dictionary()
+	dic.StringTable().Append("")
+	dic.StringTable().Append("key.a")
+
+	attr := dic.AttributeTable().AppendEmpty()
+	attr.SetKeyStrindex(1)
+	attr.Value().SetStr("value-a")
+
+	v, err := getAttributeString(dic, "key.a")
+	require.NoError(t, err)
+	require.Equal(t, "value-a", v)
+
+	_, err = getAttributeString(dic, "missing")
+	require.Error(t, err)
+	require.ErrorIs(t, err, errNotFound)
+}
+
+func TestGetAttributeStringWithPrefix(t *testing.T) {
+	t.Run("found", func(t *testing.T) {
+		dic := pprofile.NewProfiles().Dictionary()
+		dic.StringTable().Append("")
+		dic.StringTable().Append("pprof.profile.comment" + ".1")
+		dic.StringTable().Append("pprof.profile.comment" + ".0")
+
+		attr1 := dic.AttributeTable().AppendEmpty()
+		attr1.SetKeyStrindex(1)
+		attr1.Value().SetStr("comment-1")
+
+		attr0 := dic.AttributeTable().AppendEmpty()
+		attr0.SetKeyStrindex(2)
+		attr0.Value().SetStr("comment-0")
+
+		got, err := getAttributeStringWithPrefix(dic)
+		require.NoError(t, err)
+		require.Equal(t, []string{"comment-0", "comment-1"}, got)
+	})
+
+	t.Run("invalid index", func(t *testing.T) {
+		dic := pprofile.NewProfiles().Dictionary()
+		dic.StringTable().Append("")
+		dic.StringTable().Append("pprof.profile.comment" + ".x")
+
+		attr := dic.AttributeTable().AppendEmpty()
+		attr.SetKeyStrindex(1)
+		attr.Value().SetStr("comment")
+
+		_, err := getAttributeStringWithPrefix(dic)
+		require.Error(t, err)
+	})
+
+	t.Run("not found", func(t *testing.T) {
+		dic := pprofile.NewProfiles().Dictionary()
+		dic.StringTable().Append("")
+
+		_, err := getAttributeStringWithPrefix(dic)
+		require.Error(t, err)
+		require.ErrorIs(t, err, errNotFound)
+	})
 }
 
 func TestConvertMultipleSampleTypes(t *testing.T) {
@@ -472,5 +876,58 @@ func TestDefaultSampleTypeConvention(t *testing.T) {
 		for j, val := range sample.Value {
 			require.Equal(t, val, roundTrip.Sample[i].Value[j], "Round-trip should preserve original sample values order")
 		}
+	}
+}
+
+func generatePprofTestdata(t *testing.T, typ string) *profile.Profile {
+	t.Helper()
+
+	pprofile := pprof.Lookup(typ)
+	if pprofile == nil {
+		t.Fatalf("%s is not a valid pprof profile", typ)
+	}
+
+	// Generate some "load" for pprof
+	prev, current := 0, 1
+	for i := 2; i <= 42; i++ {
+		next := prev + current
+		prev = current
+		current = next
+	}
+
+	t.Logf("Fibonacci for 42 is %d", current)
+
+	tmpFile, err := os.CreateTemp(t.TempDir(), typ)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		tmpFile.Close()
+		os.Remove(tmpFile.Name())
+	})
+
+	err = pprofile.WriteTo(tmpFile, 0)
+	require.NoError(t, err)
+
+	_, err = tmpFile.Seek(0, 0)
+	require.NoError(t, err)
+
+	output, err := profile.Parse(tmpFile)
+	require.NoError(t, err)
+
+	return output
+}
+
+func TestConvertPprofToProfiles_AllTypes(t *testing.T) {
+	// cpu profiles are covered with TestConvertPprofToPprofile and
+	// the original testdata from pprof.
+	for _, typ := range []string{"goroutine", "allocs", "heap", "threadcreate", "block", "mutex"} {
+		t.Run(typ, func(t *testing.T) {
+			pprofile := generatePprofTestdata(t, typ)
+			pprofiles, err := ConvertPprofToProfiles(pprofile)
+			if err != nil {
+				t.Fatal(err)
+			}
+			_ = pprofiles
+		})
 	}
 }

--- a/pkg/translator/pprof/profiles_to_pprof.go
+++ b/pkg/translator/pprof/profiles_to_pprof.go
@@ -157,8 +157,7 @@ func convertPprofileToPprof(src *pprofile.Profiles) (*profile.Profile, error) {
 	}
 
 	var commentStrs []string
-	commentStrs, attrErr = getAttributeStringWithPrefix(src.Dictionary(),
-		string(semconv.PprofProfileCommentKey))
+	commentStrs, attrErr = getAttributeStringWithPrefix(src.Dictionary())
 	if attrErr != nil && !errors.Is(attrErr, errNotFound) {
 		return nil, attrErr
 	}
@@ -221,10 +220,11 @@ func getAttributeString(dic pprofile.ProfilesDictionary, key string) (string, er
 }
 
 // getAttributeStringWithPrefix walks the attribute_table and returns a string array
-// for all keys that start with the given prefix.
+// for all keys that start with the prefix pprof.profile.comment.
 // It returns errNotFound if the key can not be found in attribute_table.
-func getAttributeStringWithPrefix(dic pprofile.ProfilesDictionary, keyprefix string) ([]string, error) {
+func getAttributeStringWithPrefix(dic pprofile.ProfilesDictionary) ([]string, error) {
 	tmp := make(map[int]string)
+	keyprefix := string(semconv.PprofProfileCommentKey)
 
 	for _, attr := range dic.AttributeTable().All() {
 		attrKey := getStringFromIdx(dic, int(attr.KeyStrindex()))

